api/client.go

```diff
@@ -118,39 +120,17 @@ func (c *httpClient) URL(ep string, args map[string]string) *url.URL {
 	return &u
 }
 
-func (c *httpClient) Do(ctx context.Context, req *http.Request) (*http.Response, []byte, error) {
+func (c *httpClient) Do(ctx context.Context, req *http.Request) (_ *http.Response, _ []byte, err error) {
 	if ctx != nil {
 		req = req.WithContext(ctx)
 	}
 	resp, err := c.client.Do(req)
-	defer func() {
-		if resp != nil {
-			resp.Body.Close()
-		}
-	}()
-
 	if err != nil {
 		return nil, nil, err
 	}
+	defer errcapture.ExhaustClose(&err, resp.Body, "close response body")
 
-	var body []byte
-	done := make(chan struct{})
-	go func() {
-		var buf bytes.Buffer
-		_, err = buf.ReadFrom(resp.Body)
-		body = buf.Bytes()
-		close(done)
-	}()
-
-	select {
-	case <-ctx.Done():
-		<-done
-		err = resp.Body.Close()
-		if err == nil {
-			err = ctx.Err()
-		}
-	case <-done:
-	}
-
-	return resp, body, err
+	var buf bytes.Buffer
+	_, err = buf.ReadFrom(resp.Body)
```

Comment:
not having a way for `ctx` to cancel a slow read seems dangerous.

Reply:
I think we can have a context-aware cancellable reader as an internal thing and utilize it here. The functionality was previously there. 

https://pace.dev/blog/2020/02/03/context-aware-ioreader-for-golang-by-mat-ryer.html
https://benjamincongdon.me/blog/2020/04/23/Cancelable-Reads-in-Go
https://github.com/dolmen-go/contextio

Reply:
BTW (as far as I know) the only reason a `bytes.Buffer` was ever used in this function is to allow cancellation.

This is the subject of #977.